app/models/health/vaccination.rb

```diff
@@ -0,0 +1,166 @@
+
+    scope :unassigned, -> do
+      where(client_id: nil)
+    end
+
+    scope :assigned, -> do
+      where.not(client_id: nil)
+    end
+
+    scope :with_phone, -> do
+      where.not(follow_up_cell_phone: nil)
+    end
+
+    def self.csv_map(version: nil)
+      {
+        PAT_ID: :epic_patient_id,
+        MEDICAID_ID: :medicaid_id,
+        PAT_LAST_NAME: :last_name,
+        PAT_FIRST_NAME: :first_name,
+        DOB: :dob,
+        PHONE: :follow_up_cell_phone,
+        SSN: :ssn,
+        VACCINATION_DATE: :vaccinated_on,
+        VACCINATION_TYPE: :vaccination_type,
+        DEPARTMENT_NAME: :vaccinated_at,
+      }
+    end
+
+    # Called from ImportEpic
+    def self.process_new_data(values)
+      # Remove any that were removed from the source
+      remove_missing!(values)
+
+      # Import new
+      import(
+        values,
+        on_duplicate_key_update: {
+          conflict_target: conflict_key,
+          columns: [:first_name, :last_name, :ssn, :dob, :vaccinated_at, :vaccination_type, :follow_up_cell_phone],
+        },
+      )
+      propagate_to_clients
+
+      queue_sms if RailsDrivers.loaded.include?(:text_message)
+    end
+
+    def self.conflict_key
+      [:epic_patient_id, :vaccinated_on]
+    end
+
+    def self.remove_missing!(values)
+      existing_keys = pluck(*conflict_key)
+      return unless existing_keys.present?
+
+      incoming_keys = values.map { |m| m.slice(*conflict_key).values }.map { |m_id, date| [m_id, date.to_date] }
+      to_remove = existing_keys - incoming_keys
+      to_remove.each do |epic_patient_id, vaccinated_on|
+        where(epic_patient_id: epic_patient_id, vaccinated_on: vaccinated_on.to_date).destroy_all
+      end
+    end
+
+    def self.propagate_to_clients
+      # Update existing Health Emergency vaccinations if they've changed
+      assigned.preload(:he_vaccination).find_each do |vaccination|
+        vaccination.he_vaccination.vaccinated_on = vaccination.vaccinated_on
+        vaccination.he_vaccination.vaccinated_type = vaccination.clean_vaccination_type
+        vaccination.he_vaccination.vaccinated_at = vaccination.vaccinated_at
+        vaccination.he_vaccination.follow_up_cell_phone = vaccination.follow_up_cell_phone
+        vaccination.he_vaccination.save if vaccination.he_vaccination.changed?
+      end
+      # Remove any imported Health Emergency vaccinations where we've deleted the imported one
+      GrdaWarehouse::HealthEmergency::Vaccination.imported.
+        where.not(health_vaccination_id: assigned.pluck(:id)).
+        destroy_all
+
+      # Add new vaccinations
+      system_user_id = User.setup_system_user.id
+      new_vaccinations = []
+      unassigned.preload(:patient).find_each do |vaccination|
+        # Attempt to find the client based on medicaid_id
+        client_id = if vaccination.patient&.client_id.present?
+          vaccination.patient.client_id
+        else
+          all_matches = new.matching_clients(
```

Comment:
I don't really like a stateless new to access what could be a class method...  Move `matching_clients` to `self.matching_clients` and add a method with a `self.class` call?

Reply:
The `clients` method caches a lookup table of destination clients.  This was in the wrong place, so I moved it out of the loop and added a note.